Clifford+T transpiler pipeline

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

This PR (based on many discussion with @ShellyGarion and many-many-many discussion with @Cryoris) addresses #13695 by providing a transpiler pipeline for transpiling circuits into the discrete basis set of Clifford+T gates.

In many ways this is still a naive approach which misses many potential optimizations that we will consider in follows-ups. However, two goals that this aims to achieve already are:
1. Approximate non-Clifford single-qubit unitary gates using Solovay-Kitaev decomposition. This step usually requires adding many T/Tdg gates.
2. If the circuit is already a Clifford or a Clifford+T circuit, do not let the transpiler make it significantly worse.

Related to the first item, this PR is implemented on top of #14217 which provides several fixes to the Solovay-Kitaev decomposition (without which the pass very often raises an exception). Alternatively, the PR could be rebased on top of #14203 instead.

Related to the second item, the two-qubit unitary resynthesis is not currently used (in any of the transpiler stages) as it tends to replace "nice" gate sequences (consisting of few Clifford+T gates) by "very bad" gate sequences (consisting of CX and general-angle rotation gates, that, after applying Solovay-Kitaev decomposition, end up using tons of single-qubit gates). As we have just checked with @ShellyGarion, addressing #14170 does not solve the problem, and as already promised, we will explore potential improvements as follow-up.

This transpiler pipeline would also benefit from #14180 (where the definitions of standard gates with efficient Clifford+T decompositions are updated to use these decompositions), but is not based on top of that.

Functionality-wise, the default plugins for ``init``, ``translation`` and ``optimization`` stages of the transpiler pipeline now detect whether the given basis set is Clifford+T, in which case the dedicated Clifford+T plugins are used instead.

#### List of potential follow-ups.
- Resynthesize sequences of Clifford gates only using Clifford gates. Also resynthesize unitary gates that happen to be Cliffords only using Clifford gates. Maybe a dedicated Clifford resynthesis pass.
- Reduce sizes of circuits produced in the Solovay-Kitaev decomposition (SKD). There might be an opportunity to modify the SKD algorithm itself to produce shorter sequences (e.g. when a recursive approximation already satisfies the required tolerance). There seem to be opportunities to significantly simplify sequences of Clifford+T gates produced by SKD (maybe using template optimization? maybe using Rustiq? maybe a dedicated transpiler pass?). 
- Implement a Clifford+T optimization transpiler pass for 1q gates. 
- Investigate unitary synthesis algorithms into Clifford+T for two-qubit unitaries
- Implement other unitary synthesis algorithms (see #3212 for example)
- implement a general Clifford+T optimization transpiler pass
- switch to T-count metric during optimization
- produce efficient Clifford+T circuits for MCX and Adder gates when the basis set is Clifford+T
- extending the `Target` class to support `Cliffords` natively
- translate `approximation_degree` to tolerance for the Solovay-Kitaev decomposition. Adapt SKD to synthesize wrt a given tolerance.

Others ideas and pointers to relevant literature are highly welcome!

alexanderivrii committed:
experiments

alexanderivrii committed:
experiment

alexanderivrii committed:
clifford plugin

alexanderivrii committed:
optimization

alexanderivrii committed:
optimization loop

alexanderivrii committed:
remove consolidate

alexanderivrii committed:
Merge branch 'main' into discrete_transpiler_pipeline

alexanderivrii committed:
Fix _compute_rotation_axis method

The method used to return [0, 0, 0] for rotation axis when theta is very close to 0 (and yet not within 1e010).
This commit fixes this by noting that the rotation axis of an SO(3) matrix is simply the eigenvector of this
matrix corresponding to eigenvalue 1.

alexanderivrii committed:
Fixes related to SU(2) vs. SO(3)

The implemented algorithm has a global phase uncertainty of +-1, due to approximating not the
original SU(2) matrix but its projection onto SO(3). This fixes the global phase of the
computed approximation.

In addition, the product_su2 matrix in GateSequence was not correctly computed in many cases,
leading to incorrect values when using this attribute. Since this is not needed for the actual
algorithm (and spends unnecessary effort for computing it), this also completely removes this
attribute.

Co-authored-by: Almudena Carrera Vazquez <almudenacarreravazquez@hotmail.com>

alexanderivrii committed:
Fix the tests.

One of the tests was simply wrong because the reference circuit has the wrong global phase.
The tests that check the gates in the final decomposition should not assume that SGate is included and SdgGate is not.

alexanderivrii committed:
reno

alexanderivrii committed:
fix to the global phase

alexanderivrii committed:
bug fix (forgetting that gate_matrix_su2 is not SU(2) was a sequence)

alexanderivrii committed:
adding test

alexanderivrii committed:
Undoing changes to SK

alexanderivrii committed:
Merge branch 'sk_bug_fixes' into discrete_transpiler_pipeline

alexanderivrii committed:
cleanup

alexanderivrii committed:
Adding tests for discrete pass manager pipeline

alexanderivrii committed:
adding generate_discrete_passmanager to init files

alexanderivrii committed:
simplifying import

alexanderivrii committed:
another test

alexanderivrii committed:
release note

alexanderivrii committed:
Fixing tests: with the change to SK plugin, the synthesis algorithm for unitary gates that are Clifford may produce other gates

alexanderivrii committed:
Merge branch 'main' into discrete_transpiler_pipeline

alexanderivrii committed:
In SolovayKitaevSynthesis unitary synthesis plugin now storing a dictionary from the basis set to the SolovayKitaevDecomposition

alexanderivrii committed:
adding explicit test

alexanderivrii committed:
lint

alexanderivrii committed:
remove prints

alexanderivrii committed:
merge with main

alexanderivrii committed:
renaming

alexanderivrii committed:
modifying the default init/translation/optimization pass manager stage plugins to detect whether we are transpiling into a Clifford+T basis, in which we call the dedicated stage plugins. In addition, removing the function generate_clifford_t_pass_manager

alexanderivrii committed:
deleting blank line (randomly introduced at some point)

alexanderivrii committed:
change to function that checks if basis is Clifford+T

alexanderivrii committed:
make Clifford synthesis inside SolovayKitaevSynthesis plugin respect basis gates; fix tests

alexanderivrii committed:
pass over tests

alexanderivrii committed:
removing the Clifford synthesis part for now

alexanderivrii committed:
Improving is_clifford_t_basis check

alexanderivrii committed:
makung sure GateDirection pass is implied

alexanderivrii committed:
unused imports

alexanderivrii committed:
cleanup

alexanderivrii committed:
Adding Julien as co-author (and fixing a small typo)

Co-authored-by: Julien Gacon <jules.gacon@googlemail.com>

alexanderivrii committed:
lint and is_clifford_t

alexanderivrii committed:
Merge branch 'main' into discrete_transpiler_pipeline

alexanderivrii committed:
fixing pipeline to support more general Clifford+T bases; adding tests

alexanderivrii committed:
fix tests; add test with Target

alexanderivrii committed:
computing whether the basis set is clifford_t only once; caching the result

alexanderivrii committed:
cleanup; improving release notes; varying optimization_level in tests

alexanderivrii committed:
restoring ccx test

alexanderivrii committed:
removing code duplication

alexanderivrii committed:
simplifying how Clifford+T code is called

alexanderivrii committed:
updating transpiler documentation

alexanderivrii commented:
This is ready for review.

Cryoris commented on qiskit/circuit/library/standard_gates/equivalence_library.py:
```diff
@@ -1159,6 +1184,22 @@ def _cnot_rxx_decompose(plus_ry: bool = True, plus_rxx: bool = True):
+def_t.append(TdgGate(), [q[0]], [])
+def_t.append(TdgGate(), [q[0]], [])
+def_t.append(TdgGate(), [q[0]], [])
+def_t.append(TdgGate(), [q[0]], [])
```
maybe this? 😛 
```suggestion
for _ in range(7):
    def_t.append(TdgGate(), [q[0]], [])
```

Cryoris commented on qiskit/quantum_info/operators/symplectic/clifford_circuits.py:
```diff
@@ -554,5 +554,33 @@ def _append_ecr(clifford, qubit0, qubit1):
+    "ecr",
+    "dcx",
+    "t",
+    "tdg",
```
That would simplify quantum computing quite a bit 😛 
```suggestion
```

ShellyGarion replied:
I think that we should have a single place with hard-coded clifford gate names.

perhaps we should use `_BASIS_1Q` and `_BASIS_2Q` keys like is done here:
https://github.com/Qiskit/qiskit/blob/234b0852d0443b16244aba6db0a21e9029da875b/qiskit/transpiler/passes/optimization/collect_cliffords.py#L82

and if we decide to have `_CLIFFORD_GATE_NAMES` then we should use them in CollectCliffords as well.

ShellyGarion replied:
BTW, do we identify RZ gates with certain angles as Clifford? (and similarly other gates). Or do we expect SK to handle this?

alexanderivrii replied:
Haha, thanks for spotting this!

Cryoris replied:
`SolovayKitaev` should handle that case -- though I think ideally some kind of basis translator would be able to figure this out, which should be less costly than SK

Cryoris commented on qiskit/transpiler/__init__.py:
```diff
@@ -673,6 +673,9 @@
 towards the ISA.
 
+For a Clifford+T basis set, the single-qubit rotation gates are approximated using the
+Solovay-Kitaev algorithm.
```
Maybe add a hyperref to the `SolovayKitaev` class used for this? 🙂

Cryoris commented on qiskit/transpiler/passmanager_config.py:
```diff
@@ -95,6 +95,9 @@ def __init__(
         self.qubits_initially_zero = qubits_initially_zero
+        # Stores whether the basis gates are Clifford+T,
+        # in which case we use stage manager plugins adapted to Clifford+T.
+        self._is_clifford_t = False
```
For now this works, but if we end up with different basis gate sets for FTQC (such as PBC) then we'll need to make this more extensible by maybe determining the basis gate mode and having separate pass managers for these

alexanderivrii replied:
Agreed!

Cryoris commented on qiskit/transpiler/preset_passmanagers/builtin_plugins.py:
```diff
@@ -496,6 +507,13 @@ class OptimizationPassManager(PassManagerStagePlugin):
 
     def pass_manager(self, pass_manager_config, optimization_level=None) -> PassManager:
         """Build pass manager for optimization stage."""
+
```
IMO it would be cleaner to have a `DefaultOptimizationPassManager` too and just use `OptimizationPassManager` as a dispatch. But we can also leave this as is for now and maybe reconsider if we have multiple basis gate modes.

alexanderivrii replied:
This is the default plugin for the optimization stage, see https://github.com/Qiskit/qiskit/blob/234b0852d0443b16244aba6db0a21e9029da875b/pyproject.toml#L147

I agree that it would be nice to change the name to reflect this and to be consistent with the naming of other default plugins, but thought that I should not do this change in this PR.

Cryoris commented on releasenotes/notes/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -0,0 +1,41 @@
+
+    It is important to note that the specified Clifford+T basis gate set should be
+    universal, or else transpilation might not suceed. While the gate set 
+    ``["h", "t", "tdg"]``or even ``["h", "t"]`` is sufficient for universality,
```
```suggestion
    universal, or else transpilation might not succeed. While the gate set 
    ``["h", "t", "tdg"]`` or even ``["h", "t"]`` is sufficient for universality,
```

alexanderivrii replied:
I am optimizing, one letter at a time 😄

Cryoris reviewed (commented):
This looks and feels already quite nice (I might be biased lol). I left some comments below, and also:
* I think it would be really good if we can add optimizations that merge consecutive T gates into S or Z into 2.1 as well, since it looks like it could reduce T count quite a bit
* Right now SK seems to take precedence over possible basis translations: e.g. transpiling `SX` into `T Tdg S Sdg H` could give the Clifford-decomp of the basis translator `Sdg H Sdg` but it yields `H T T H` right now

ShellyGarion commented on qiskit/circuit/library/standard_gates/equivalence_library.py:
```diff
@@ -797,6 +797,17 @@ def _cnot_rxx_decompose(plus_ry: bool = True, plus_rxx: bool = True):
 def_s.append(U1Gate(pi / 2), [q[0]], [])
 _sel.add_equivalence(SGate(), def_s)
 
+# SGate
```
I think that for ZGate we already have the identity: Z = S S but we don't have Z = Sdg Sdg. Is it worth to add it?

Will these identities imply that T T T T (or Tdg Tdg Tdg Tdg) will be replaced by Z ?

alexanderivrii replied:
That's a good point. For now, I have added the minimal amount of equivalences to make this PR work, but we are definitely missing some of the important equivalences. Going forward, I think that by far the best strategy is to include @jakelishman's `BasisConstructor` pass (together with the corresponding library), Jake did a really nice job at thinking through the hierarchy of equivalences that we should add to the circuit library.

Cryoris replied:
I don't think these identities enable `T T T T -> Z` since the basis translator only contains `1 -> N` relations and cannot merge gates. Merging `T` sequences is planned as follow up (@alexanderivrii already has a draft), hopefully for 2.1.

ShellyGarion commented on qiskit/transpiler/preset_passmanagers/builtin_plugins.py:
```diff
@@ -1003,3 +1021,63 @@ def _get_trial_count(default_trials=5):
+        optimization = PassManager()
+        if optimization_level != 0:
+            _depth_check = [Depth(recurse=True), FixedPoint("depth")]
+            _size_check = [Size(recurse=True), FixedPoint("size")]
```
is there an optimization for specific gate of type of gates?
I thought that usually we aim to optimize 2q gates (CX, CZ, RZZ). and in this case we should optimize non-clifford (T, Tdg) gates.

alexanderivrii replied:
I agree, but again I think it's something we should pursue in a follow-up. Anyhow, for the current set of Clifford+T optimization passes, all of the metrics are the same.

ShellyGarion commented on releasenotes/notes/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -0,0 +1,41 @@
+      from qiskit.transpiler.preset_passmanagers import generate_preset_pass_manager
+      from qiskit.quantum_info import get_clifford_gate_names
+
+      basis_gates = get_clifford_gate_names() + ["t", "tdg"]
```
perhaps in the example we would like to fix the basis gates, e.g. `[cx, h, t, tdg]`

ShellyGarion replied:
btw, do we assume that the clifford gates are hard-coded? can the user define their own clifford gate and provide it (say as clifford tableaux)?

alexanderivrii replied:
One of the points of the example is to use `get_clifford_gate_names` to get the list of all standard Clifford gate names  in Qiskit. Most users will probably want to include as many Clifford gates as possible, and this example illustrates how this can be done. Of course, the users can specify their own gate lists, such as ``["h", "t", "tdg"]``or even ``["h", "t"]``, please see the last paragraph in the release notes.

So far in all of the transpiler passes ``basis_gates`` is a list of strings, and we probably do not want to generalize this. However, your comment is very relevant for extending the `Target` class for supporting Clifford+T gates in a form that is hopefully better than just listing all the Clifford gate names.

ShellyGarion commented on releasenotes/notes/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -0,0 +1,41 @@
+      basis_gates = get_clifford_gate_names() + ["t", "tdg"]
+      pm = generate_preset_pass_manager(basis_gates=basis_gates)
+
+      qc = QuantumCircuit(1)
```
perhaps it's worth to provide a circuit with 2-3 qubits? otherwise, the user could just use SK plugin

alexanderivrii replied:
There are more examples available as tests. 😄

ShellyGarion replied:
I know that there are tests, still the user would like to know what's new in here compared to using SK directly, and I think that essentially it's the fact that you can do it for several qubits.

alexanderivrii replied:
Sure, added a slightly larger example in [a6799f1](https://github.com/Qiskit/qiskit/pull/14225/commits/a6799f1d8e0966afbc5cc69734b5c79391cd312c).

ShellyGarion commented on test/python/transpiler/test_clifford_t_passmanager.py:
```diff
@@ -0,0 +1,233 @@
+        )
+        transpiled = pm.run(qc)
+        self.assertLessEqual(set(transpiled.count_ops()), set(basis_gates))
+
```
in this case, one should use only clifford gates for the transpilation?

alexanderivrii replied:
That would be true with the Clifford plugin I mentioned somewhere, but not yet. For efficiency, we are currently running the Solovay-Kitaev synthesis algorithm with the basis set ``["h", "t", "tdg"]``, so even pure Clifford gates would be transpiled into this basis. Note that you could run the SK algorithm with a much larger set of basis gates, like all Clifford gates + T/Tdg, however at least the python implementation becomes significantly slower. This is something to revisit when the SK algorithm is ported to Rust.

ShellyGarion commented on qiskit/transpiler/preset_passmanagers/builtin_plugins.py:
```diff
@@ -1003,3 +1021,63 @@ def _get_trial_count(default_trials=5):
     return default_trials
+
+
+class CliffordTOptimizationPassManager(PassManagerStagePlugin):
```
Do you use somewhere the `CollectClliffords` transpiler pass? I think it can collect parametrized gates that are clifford only for certain parameters.

alexanderivrii replied:
Good point. I am not using this yet. Again this relates to the point that in the future we want to have Clifford and Clifford+T circuit resynthesis methods.

ShellyGarion commented on test/python/transpiler/test_clifford_t_passmanager.py:
```diff
@@ -0,0 +1,233 @@
+            basis_gates=basis_gates, optimization_level=optimization_level
+        )
+        transpiled = pm.run(qc)
+        self.assertLessEqual(set(transpiled.count_ops()), set(basis_gates))
```
in this case, one should use only clifford gates for the transpilation?

alexanderivrii replied:
Same answer as before. :)

ShellyGarion commented on test/python/transpiler/test_clifford_t_passmanager.py:
```diff
@@ -0,0 +1,233 @@
+        requiring the usage of the Solovay-Kitaev decomposition.
+        """
+        qc = QuantumCircuit(1)
+        qc.rx(0.8, 0)
```
if the angle is a multiplication of pi/2, does it know to use only clifford gates?
if the angle is a multiplication of pi/4, does it know to use only clifford and t/tdg gates?

alexanderivrii replied:
For the pi/4 angle, if the basis set is Clifford+T, the flow by definition will only use Clifford+T/Tdg gates. But you are correct, the SK algorithm will be exact and produce a very short sequence of gates. I am adding a test for that. I don't want to add a similar test for pi/2 angle before we can guarantee (again using the future Clifford resynthesis plugin) that there will be only Clifford gates in the result.

ShellyGarion commented on test/python/transpiler/test_clifford_t_passmanager.py:
```diff
@@ -0,0 +1,233 @@
+
+        # Should get the efficient decomposition of the Toffoli gates into Clifford+T.
+        self.assertEqual(transpiled.count_ops(), {"cx": 6, "t": 4, "tdg": 3, "h": 2})
+
```
these are very nice examples! perhaps we can consider adding more advanced ones, like c3x, c4x, mcx or certain adders?

alexanderivrii replied:
As I have already mentioned somewhere, producing _efficient_ circuits is something we want to do, but this requires a bit more work that we want to do in a follow-up. First, we need to make sure that all the standard definitions for gates like c3x, c4x only use Clifford+T gates (I have an open PR on that somewhere), and then we need to update HighLevelSynthesis plugins to always choose Clifford+T synthesis algorithms when available.

ShellyGarion reviewed (commented):
Thanks for this important work. I have a few comments and questions.

alexanderivrii committed:
addressing review comments

alexanderivrii commented:
For some reason there are comments to which I can't seem to comment back, so I will include my replies here as I go through them. 

@Cryoris:
> I think it would be really good if we can add optimizations that merge consecutive T gates into S or Z into 2.1 as well, since it looks like it could reduce T count quite a bit

I am planning to submit a follow-up PR shortly. 

> Right now SK seems to take precedence over possible basis translations: e.g. transpiling SX into T Tdg S Sdg H could give the Clifford-decomp of the basis translator Sdg H Sdg but it yields H T T H right now

I agree and this is something I definitely want to improve in the future. In fact, in the previous version of this PR I have modified the Solovay-Kitaev unitary synthesis plugin to detect when a 1q-gate is Clifford, in which case to use a Clifford resynthesis algorithm rather than Solovay-Kitaev. However, I have later removed this optimization since the resynthesized circuit did not adhere to the basis set passed to Solovay-Kitaev. We should think what is the best way to include this optimization (but in a follow-up): another unitary synthesis plugin? standalone pass?

@ShellyGarion:

> perhaps we should use _BASIS_1Q and _BASIS_2Q keys like is done here:

This is exactly how I have implemented this at first, before I have found out that some of the gate names in `_BASIS_1Q` and `_BASIS_2Q` are not standard gate names, e.g., `i`, `iden`, `v`, `w`.

> and if we decide to have _CLIFFORD_GATE_NAMES then we should use them in CollectCliffords as well.

I can do this, if you want me to, though I feel that this change is somewhat unrelated to this PR.

> BTW, do we identify RZ gates with certain angles as Clifford? (and similarly other gates). Or do we expect SK to handle this?

Yes. Or the future Clifford resynthesis plugin I mentioned earlier.

alexanderivrii commented:
@ShellyGarion, I have added your comments on potential followups in the PR's description. If you think that some ideas are missing, please update the description as you see fit.

alexanderivrii committed:
adding a larger example to releasenotes

ShellyGarion reviewed (approved):
LGTM :)

Cryoris commented on releasenotes/notes/add-clifford-t-transpiler-pipeline-e640b5ba64eab13a.yaml:
```diff
@@ -32,10 +32,28 @@ features_transpiler:
     decomposing Clifford gates into non-Clifford gates, which might not be the
     desired behavior).
+
+    Here is an additional slightly larger example::
```
2 examples are a bit overkill IMO, I think the top one is good enough (you could add a second gate there if you wanted or something)

alexanderivrii replied:
Shelly was specifically asking for an example over multiple qubits 😄. But actually the two examples are slightly different: the first uses `get_clifford_get_names`, and the second a custom Clifford+T basis.

Cryoris replied:
The release manager shall rule 😄
